prometheus-to-sd/config/dynamic_source.go

```diff
@@ -71,33 +87,81 @@ func createOptionsForPodSelection(nodeName string, sources map[string]url.URL) v
 	for key := range sources {
 		nameList += key + ","
 	}
-	labelSelector := fmt.Sprintf("%s in (%s)", nameLabel, nameList)
-	return v1.ListOptions{
+	labelSelector := fmt.Sprintf("%s in (%s)", componentNameLabel, nameList)
+	listOptions := v1.ListOptions{
 		FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodeName),
 		LabelSelector: labelSelector,
 	}
+	glog.Infof("Filters for dynamic components: %+v", listOptions)
+	return listOptions
 }
 
-func getConfigsFromPods(pods []core.Pod, sources map[string]url.URL) []SourceConfig {
-	var sourceConfigs []SourceConfig
-	for _, pod := range pods {
-		componentName := pod.Labels[nameLabel]
-		source, _ := sources[componentName]
-		podConfig := PodConfig{
-			PodId:       pod.Name,
-			NamespaceId: pod.Namespace,
-		}
-		sourceConfig, err := mapToSourceConfig(componentName, source, pod.Status.PodIP, podConfig)
-		if err != nil {
-			glog.Warning("could not create source config for pod %s: %v", pod.Name, err)
-		}
-		sourceConfigs = append(sourceConfigs, *sourceConfig)
+func getConfigFromPod(pod core.Pod, sources map[string]url.URL) (*SourceConfig, error) {
+	if pod.Status.PodIP == "" {
+		return nil, errors.New("pod ip is missing")
+	}
+	componentName := pod.Labels[componentNameLabel]
+	source, _ := sources[componentName]
+	podConfig := PodConfig{
+		PodId:       pod.Name,
+		NamespaceId: pod.Namespace,
 	}
-	return sourceConfigs
+	return mapToSourceConfig(componentName, source, pod.Status.PodIP, podConfig)
 }
 
 func mapToSourceConfig(componentName string, url url.URL, ip string, podConfig PodConfig) (*SourceConfig, error) {
 	port := url.Port()
 	whitelisted := url.Query().Get("whitelisted")
 	return newSourceConfig(componentName, ip, port, url.Path, whitelisted, podConfig)
 }
+
+func getSourceConfigChangeChannel(kubeApi clientset.Interface, sources map[string]url.URL, podOptions v1.ListOptions) (<-chan SourceConfigChange, error) {
+	watchInterface, err := kubeApi.CoreV1().Pods(podNamespace).Watch(podOptions)
+	if err != nil {
+		return nil, err
+	}
+	sourceConfigChannel := make(chan SourceConfigChange)
+	eventChannel := watchInterface.ResultChan()
```

Comment:
Can you defer closing this channel in gorouting. Even in infinite loop is good to free resources that you take, so someone other when rewriting code will not skip it.